apis/postgresql/v1alpha1/postgresqluserrole_types.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"github.com/easymile/postgresql-operator/apis/postgresql/common"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+type PostgresqlUserRolePrivilege struct {
+	// User privileges
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=OWNER;WRITER;READER
+	Privilege PrivilegesSpecEnum `json:"privilege"`
+	// Postgresql Database
+	// +required
+	// +kubebuilder:validation:Required
+	Database *common.CRLink `json:"database"`
+	// Generated secret name prefix
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	GeneratedSecretName string `json:"generatedSecretName"`
+}
+
+type ModeEnum string
+
+const ProvidedMode ModeEnum = "PROVIDED"
+const ManagedMode ModeEnum = "MANAGED"
+
+// PostgresqlUserRoleSpec defines the desired state of PostgresqlUserRole.
+type PostgresqlUserRoleSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// User mode
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum=PROVIDED;MANAGED
+	Mode ModeEnum `json:"mode,omitempty"`
+	// Privileges
+	// +required
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Min=1
+	Privileges []*PostgresqlUserRolePrivilege `json:"privileges"`
+	// User role prefix
+	// +optional
+	RolePrefix string `json:"rolePrefix,omitempty"`
```

Comment:
This is ambiguous to have required fields only in one mode.

It seems the only solution is validation rules, but only available in k8s 1.25 (in beta...)
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules